### PR TITLE
Webcfg Metadata TR181 DML and WEBPA Support

### PR DESCRIPTION
**Pull request by guruchandru**

### CODE REVIEW

File: source/WebConfig/cosa_webconfig_dml.c [hunk 1]
```diff
@@ -163,7 +173,7 @@ X_RDK_WebConfig_GetParamStringValue
         }
         if( AnscEqualString(ParamName, "Data", TRUE))
 	    {
-                WebcfgDebug("[%s] at [%d]parameter '%s'\n",__FUNCTION__,__LINE__, ParamName);
+                WebcfgDebug(("[%s] at [%d]parameter '%s'\n",__FUNCTION__,__LINE__, ParamName));
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:176]: Double bracket is not required for webcfg logging

File: source/WebConfig/cosa_webconfig_dml.c [hunk 2]
```diff
@@ -188,9 +198,63 @@ X_RDK_WebConfig_GetParamStringValue
 	        }
                 return 0;
         }
+        if( AnscEqualString(ParamName, "SupportedDocs", TRUE))
+	    {
+                WebcfgInfo(("[%s] at [%d]parameter '%s'\n",__FUNCTION__,__LINE__, ParamName));
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:203]: Make this debug log. Double bracket is not required for webcfg logging

File: source/WebConfig/cosa_webconfig_dml.c [hunk 3]
```diff
@@ -188,9 +198,63 @@ X_RDK_WebConfig_GetParamStringValue
...
+
+			 if(*pUlSize <= strlen(docValue))
+                         {
+			     *pUlSize = strlen(docValue) + 1;
+			     return 1;
+		         }
+		         /* collect value */
+		         AnscCopyString(pValue, docValue);
+                         WebcfgDebug("The pValue is %s\n",pValue);
+		        /* if(docValue != NULL)
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:220]: cleanup code.

File: source/WebConfig/cosa_webconfig_dml.c [hunk 4]
```diff
@@ -188,9 +198,63 @@ X_RDK_WebConfig_GetParamStringValue
...
+		         {
+			     free(docValue);
+			     docValue = NULL;
+		         }*/
+	        }
+                return 0;
+        }
+        if( AnscEqualString(ParamName, "SupportedSchemaVersion", TRUE))
+	    {
+                WebcfgInfo(("[%s] at [%d]parameter '%s'\n",__FUNCTION__,__LINE__, ParamName));
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:230]: make it to debug and remove double bracket

File: source/WebConfig/cosa_webconfig_dml.c [hunk 5]
```diff
@@ -188,9 +198,63 @@ X_RDK_WebConfig_GetParamStringValue
...
+
+			 if(*pUlSize <= strlen(versionValue))
+                         {
+			     *pUlSize = strlen(versionValue) + 1;
+			     return 1;
+		         }
+		         /* collect value */
+		         AnscCopyString(pValue, versionValue);
+                         WebcfgDebug("The pValue is %s\n",pValue);
+		     /*    if(versionValue != NULL)
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:247]: check if free is required, else remove and cleanup this code.

File: source/WebConfig/cosa_webconfig_dml.c [hunk 6]
```diff
@@ -188,9 +198,63 @@ X_RDK_WebConfig_GetParamStringValue
...
+			     free(versionValue);
+			     versionValue = NULL;
+		         }*/
+	        }
+		return 0;
+        }
 
 	WebcfgDebug("------- %s ----- EXIT ----\n",__FUNCTION__);
- 	WebcfgError("Unsupported parameter '%s'\n", ParamName);
+ 	WebcfgError(("Unsupported parameter '%s'\n", ParamName));
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:257]: Remove double bracket

File: source/WebConfig/cosa_webconfig_internal.c [hunk 7]
```diff
@@ -316,6 +319,40 @@ int getWebConfigParameterValues(char **parameterNames, int paramCount, int *val_
...
 			      }
+                            else if((strcmp(parameterNames[i], WEBCONFIG_PARAM_SUPPORTED_DOCS) == 0) && (RFC_ENABLE == true))
+                            {
+                                  WebcfgDebug("B4 Get_Webconfig_Supported_Docs\n");
+                                  char* docvalue =  getsupportedDocs();
+                                  paramVal[k]->parameterName = strndup(WEBCONFIG_PARAM_SUPPORTED_DOCS, MAX_PARAMETERNAME_LEN);
+                                  if( (docvalue != NULL) && strlen(docvalue) >0 )
+                                  {
+                                      WebcfgDebug("Webpa get : Datafetched %s\n", docvalue);
+                                      paramVal[k]->parameterValue = strdup(docvalue);
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_internal.c:330]: Would suggest to make either strdup or strndup to make it uniform, as strdup is being used in if case and strndup in else case. Check all cases.

**Review by sadhyama — CHANGES_REQUESTED** [review]

**Review by shilpa24balaji — APPROVED** [review]

File: source/WebConfig/cosa_webconfig_dml.c [hunk 8]
```diff
@@ -126,6 +126,16 @@ X_RDK_WebConfig_SetParamStringValue
 			WebcfgInfo("Data set is Not supported\n");
             return TRUE;
 		}
+		if( AnscEqualString(ParamName, "SupportedDocs", TRUE))
+                {
+			WebcfgInfo("SupportedDocs set is Not supported\n");
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:131]: We can make this as debug log as dmcli will return CCSP_ERR_NOT_WRITABLE.

File: source/WebConfig/cosa_webconfig_dml.c [hunk 9]
```diff
@@ -126,6 +126,16 @@ X_RDK_WebConfig_SetParamStringValue
...
+                        return TRUE;
+                }
+                if( AnscEqualString(ParamName, "SupportedSchemaVersion", TRUE))
+                {
+			WebcfgInfo("SupportedSchemaVersion set is Not supported\n");
```
Comment (sadhyama) [inline source/WebConfig/cosa_webconfig_dml.c:136]: Make this as debug log as dmcli will return CCSP_ERR_NOT_WRITABLE.

**Review by sadhyama — APPROVED** [review]

**Pull request merged**